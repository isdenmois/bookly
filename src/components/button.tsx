import React from 'react';
import { TouchableOpacity, View, ViewStyle, TextStyle, Platform } from 'react-native';
import { DynamicStyleSheet, useDynamicValue } from 'react-native-dynamic';
import classnames from 'rn-classnames';
import { dynamicColor, boldText } from 'types/colors';
import { TextM } from 'components/text';

interface Props {
  label: string;
  icon?: any;
  thin?: boolean;
  disabled?: boolean;
  bordered?: boolean;
  style?: ViewStyle | ViewStyle[];
  textStyle?: TextStyle;
  testID?: string;
  onPress?: () => void;
}

export function Button(props: Props) {
  const s = useDynamicValue(ds);
  const cn = classnames(s);
  const { icon, label, thin, disabled, bordered, textStyle } = props;
  const viewStyles = [...cn('container', { thin, disabled, bordered })].concat(props.style);
  const textStyles = [s.text, ...cn(icon && { icon: !thin, iconThin: thin }, { inversed: bordered }), textStyle];
  const Component: any = disabled ? View : TouchableOpacity;

  return (
    <Component testID={props.testID} onPress={!disabled && props.onPress} style={viewStyles}>
      {!!icon && icon}
      <TextM style={textStyles}>{label}</TextM>
    </Component>
  );
}

const ds = new DynamicStyleSheet({
  container: {
    flexDirection: 'row',
    backgroundColor: dynamicColor.Primary,
    paddingHorizontal: 25,
    paddingVertical: 10,
    borderRadius: 20,
    alignItems: 'center',
    justifyContent: 'center',
  } as ViewStyle,
  icon: {
    marginLeft: 10,
  } as ViewStyle,
  iconThin: {
    marginLeft: 6,
  } as ViewStyle,
  thin: {
    paddingHorizontal: 10,
    paddingVertical: 5,
    borderRadius: 15,
  } as ViewStyle,
  disabled: {
    opacity: 0.4,
  } as ViewStyle,
  bordered: {
    backgroundColor: 'transparent',
    borderWidth: 1,
    borderColor: dynamicColor.Primary,
  } as ViewStyle,
  text: {
    color: dynamicColor.PrimaryTextInverse,
    textAlign: 'center',
    lineHeight: 19,
    ...boldText,
  } as TextStyle,
  inversed: {
    color: dynamicColor.Primary,
  } as TextStyle,
});
