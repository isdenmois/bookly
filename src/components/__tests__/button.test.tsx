import React from 'react';
import { create } from 'react-test-renderer';
import { Text } from 'react-native';
import { Button } from '../button';

describe('Button', () => {
  it('renders button', () => {
    const tree = create(<Button label='Superb' />);

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders icon', () => {
    const tree = create(<Button label='Test' icon={<Text />} />);

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders disabled', () => {
    const tree = create(<Button label="You won't click me" disabled />);

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
