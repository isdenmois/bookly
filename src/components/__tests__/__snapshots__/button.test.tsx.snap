// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders button 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#57B5AC",
      "borderRadius": 20,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 25,
      "paddingVertical": 10,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 16,
        },
        Array [
          Object {
            "color": "#FFF",
            "fontFamily": "sans-serif-medium",
            "lineHeight": 19,
            "textAlign": "center",
          },
          undefined,
        ],
      ]
    }
  >
    Superb
  </Text>
</View>
`;

exports[`Button renders disabled 1`] = `
<View
  onPress={false}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#57B5AC",
        "borderRadius": 20,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 25,
        "paddingVertical": 10,
      },
      Object {
        "opacity": 0.4,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 16,
        },
        Array [
          Object {
            "color": "#FFF",
            "fontFamily": "sans-serif-medium",
            "lineHeight": 19,
            "textAlign": "center",
          },
          undefined,
        ],
      ]
    }
  >
    You won't click me
  </Text>
</View>
`;

exports[`Button renders icon 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#57B5AC",
      "borderRadius": 20,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 25,
      "paddingVertical": 10,
    }
  }
>
  <Text />
  <Text
    style={
      Array [
        Object {
          "fontSize": 16,
        },
        Array [
          Object {
            "color": "#FFF",
            "fontFamily": "sans-serif-medium",
            "lineHeight": 19,
            "textAlign": "center",
          },
          Object {
            "marginLeft": 10,
          },
          undefined,
        ],
      ]
    }
  >
    Test
  </Text>
</View>
`;
